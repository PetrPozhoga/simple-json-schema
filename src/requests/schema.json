{
  "title": "A registration form",
  "description": "A simple form example.",
  "type": "object",
  "required": [
    "firstName",
    "lastName",
    "termsOfUse"
  ],
  "properties": {
    "firstName": {
      "type": "string",
      "title": "First name",
      "value": "Chuck",
      "autoFocus": true
    },
    "lastName": {
      "type": "string",
      "title": "Last name"
    },
    "age": {
      "type": "integer",
      "title": "Age",
      "description": "(earthian year)"
    },
    "bio": {
      "type": "string",
      "title": "Bio",
      "isTextarea": true
    },
    "password": {
      "type": "password",
      "title": "Password",
      "minLength": 3,
      "hint": "Hint: Make it strong!"
    },
    "termsOfUse": {
      "type": "checkbox",
      "title": "You agree terms of use"
    }
  }
}