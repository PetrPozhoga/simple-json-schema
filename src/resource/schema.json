{
  "title": "A registration form",
  "description": "A simple form example.",
  "required": [
    "firstName",
    "lastName",
    "password",
    "termsOfUse"
  ],
  "properties": {
    "firstName": {
      "type": "string",
      "title": "First name",
      "value": "Chuck",
      "autoFocus": true
    },
    "lastName": {
      "type": "string",
      "title": "Last name"
    },
    "age": {
      "type": "integer",
      "title": "Age",
      "description": "(earthian year)"
    },
    "bio": {
      "type": "string",
      "title": "Bio",
      "isTextarea": true
    },
    "confirmPassword": {
      "type": "confirmPassword",
      "title": "Confirm password"
    },
    "password": {
      "type": "password",
      "title": "Password",
      "minLength": "3",
      "hint": "Hint: Make it strong!"
    },
    "termsOfUse": {
      "type": "checkbox",
      "value": true,
      "title": "You agree terms of use"
    }
  },
  "submitValue": "Send"
}
