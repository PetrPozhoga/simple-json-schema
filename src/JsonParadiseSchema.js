import React from 'react'
import PropTypes from 'prop-types'
import './JsonParadiseSchema.scss'
import Form from "./components/Form/Form"
import { useSchema } from './hooks/schema.hook'
import { useValidation } from './hooks/validation.hook'
import { useForm } from "./hooks/form.hook"

function JsonParadiseSchema({ schema, onSubmit }) {

  const {
    title,
    description,
    required,
    properties,
    setProperties,
    submitValue,
    initialState,
  } = useSchema(schema)

  const { fieldValidation } = useValidation({setProperties, required})
  const { changeFields, sendForm } = useForm({onSubmit, initialState, properties, setProperties, fieldValidation})

  return (
    <div className="react-simple-json-schema">
      <div className='container'>
        { title ? <h1>{ title }</h1> : null }
        { description ? <h2>{ description }</h2> : null }
        { !!Object.keys(properties).length &&
        <Form changeFields={ changeFields } submitValue={ submitValue } sendForm={ sendForm }
              required={ required } properties={ properties }/> }
      </div>
    </div>
  )
}

JsonParadiseSchema.propTypes = {
  schema: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired,
}

export default JsonParadiseSchema
