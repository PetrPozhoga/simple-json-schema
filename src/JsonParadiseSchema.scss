@import "styles/variables";

.react-simple-json-schema {
  padding: 0;
  line-height: 18px;
  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
    box-shadow: none;
    border: none;
    outline: none;
    text-decoration: none;
    list-style: none;
    border-radius: 0;
  }
  .container {
    max-width: 500px;
    margin: 50px auto;
    padding: 0 20px;
    & > h1 {
      font-size: 22px;
      line-height: 18px;
      margin-bottom: 20px;
      color: #111;
      padding-bottom: 10px;
      border-bottom: 1.2px solid $primary-border-bg;
    }

    & > h2 {
      font-size: 18px;
      line-height: 18px;
      color: #111;
      margin-bottom: 15px;
      font-weight: 400;
    }
  }
  .primary-btn {
    background: #39A2F1;
    color: #fff;
    opacity: .8;
    display: inline-block;
    -webkit-appearance: none;
    padding: 0 15px;
    text-align: center;
    height: 35px;
    line-height: 35px;
    border-radius: 5px;
    font-size: 16px;
    cursor: pointer;
    &:hover{
      opacity: 1;
    }
  }
}