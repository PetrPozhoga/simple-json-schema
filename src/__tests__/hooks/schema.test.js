import React from 'react'
import { useSchema } from "../../hooks/schema.hook"
import schema from '../../resource/schema'
import { testHook } from "../../utils/testingComponent"
import { mount } from 'enzyme'
import JsonParadiseSchema from "../../JsonParadiseSchema"

jest.mock('../../components/Form/Form', () => () => null)

describe('describe schema', () => {

  let schemaObj = {}

  beforeEach(() => {
    testHook(() => {
      schemaObj = useSchema(schema)
    })
  })

  it('test returned value without properties, setProperties, initialState', () => {

    expect(schema).toMatchSnapshot()

    const deleteObjKey = (list, obj) => list.forEach(key => delete obj[ key ])
    deleteObjKey([ 'properties', 'setProperties', 'initialState' ], schemaObj)

    Object.keys(schemaObj).forEach(key => expect(schemaObj[ key ]).toEqual(schema[ key ]))
  })

  it('test properties (set default /key=value/ if key is undefined)', () => {
    const { properties } = schemaObj
    Object.keys(properties).forEach(key => {
      let schemaProperties = JSON.parse(JSON.stringify(schema.properties))
      schemaProperties[ key ].value = schemaProperties[ key ].value || ''
      expect(properties[ key ]).toEqual(schemaProperties[ key ])
    })
  })
})

describe('describe schema error', () => {

  it('throw if Object keys initialState length > 0 and schema object keys length = 0', () => {
    spyOn(console, 'error')
    let component = mount(<JsonParadiseSchema schema={ schema } onSubmit={ () => {} }/>)
    expect(() => component.setProps({
      schema: {},
      onSubmit: () => {}
    })).toThrow("Failed prop type: The prop `properties` is marked as required in `N`, but its value is not `object`.")
  })

  it('return if schema object keys length = 0', () => {
    let schemaObj = {}

    testHook(() => {
      schemaObj = useSchema({})
    })

    const { description, initialState, properties, required, submitValue, title } = schemaObj

    expect(description).toBe('')
    expect(initialState).toEqual({})
    expect(properties).toEqual({})
    expect(required).toEqual([])
    expect(submitValue).toBe("")
    expect(title).toBe("")
  })

  it('throw when properties for value not set', () => {
    spyOn(console, 'error')

    expect(
      () => testHook(() => {
        schema.properties = {}
        useSchema(schema)
      })
    ).toThrow('Failed prop type: The prop `properties` is marked as required in `N`, but its value is not `object`.')
  })
})
