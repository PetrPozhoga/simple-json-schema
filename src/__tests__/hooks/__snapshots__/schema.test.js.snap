// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe schema test returned value without properties, setProperties, initialState 1`] = `
Object {
  "description": "A simple form example.",
  "properties": Object {
    "age": Object {
      "description": "(earthian year)",
      "title": "Age",
      "type": "integer",
    },
    "bio": Object {
      "isTextarea": true,
      "title": "Bio",
      "type": "string",
    },
    "firstName": Object {
      "autoFocus": true,
      "title": "First name",
      "type": "string",
      "value": "Chuck",
    },
    "lastName": Object {
      "title": "Last name",
      "type": "string",
    },
    "password": Object {
      "hint": "Hint: Make it strong!",
      "minLength": "3",
      "title": "Password",
      "type": "password",
    },
    "termsOfUse": Object {
      "title": "You agree terms of use",
      "type": "checkbox",
      "value": true,
    },
  },
  "required": Array [
    "firstName",
    "lastName",
    "password",
    "termsOfUse",
  ],
  "submitValue": "Send",
  "title": "A registration form",
}
`;
